hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/InvokeOnPartitions.java

```diff
@@ -16,137 +16,212 @@
  */
 final class InvokeOnPartitions {
 
     private static final int TRY_COUNT = 10;
     private static final int TRY_PAUSE_MILLIS = 300;
 
-    private OperationServiceImpl operationService;
+    private static final Object NULL_RESULT = new Object() {
+        @Override
+        public String toString() {
+            return "NULL_RESULT";
+        }
+    };
+
+    private final OperationServiceImpl operationService;
     private final String serviceName;
     private final OperationFactory operationFactory;
     private final Map<Address, List<Integer>> memberPartitions;
-    private final Map<Address, Future> futures;
-    private final Map<Integer, Object> partitionResults;
+    private final AtomicReferenceArray<Object> partitionResults;
+    private final AtomicInteger latch;
+    private volatile ExecutionCallback<Map<Integer, Object>> callback;
+    private final SimpleCompletableFuture future;
 
     InvokeOnPartitions(OperationServiceImpl operationService, String serviceName, OperationFactory operationFactory,
                        Map<Address, List<Integer>> memberPartitions) {
         this.operationService = operationService;
         this.serviceName = serviceName;
         this.operationFactory = operationFactory;
         this.memberPartitions = memberPartitions;
-        this.futures = createHashMap(memberPartitions.size());
         int partitionCount = operationService.nodeEngine.getPartitionService().getPartitionCount();
-        this.partitionResults = createHashMap(partitionCount);
+        // this is the total number of partitions for which we actually have operation
+        int actualPartitionCount = 0;
+        for (List<Integer> mp : memberPartitions.values()) {
+            actualPartitionCount += mp.size();
+        }
+        this.partitionResults = new AtomicReferenceArray<Object>(partitionCount);
+        this.latch = new AtomicInteger(actualPartitionCount);
+        this.future = new SimpleCompletableFuture(operationService.nodeEngine);
     }
 
     /**
      * Executes all the operations on the partitions.
      */
     <T> Map<Integer, T> invoke() throws Exception {
-        ensureNotCallingFromPartitionOperationThread();
+        return this.<T>invokeAsync(null).get();
+    }
 
+    /**
+     * Executes all the operations on the partitions.
+     */
+    @SuppressWarnings("unchecked")
+    <T> ICompletableFuture<Map<Integer, T>> invokeAsync(ExecutionCallback<Map<Integer, T>> callback) {
+        this.callback = (ExecutionCallback<Map<Integer, Object>>) ((ExecutionCallback) callback);
+        ensureNotCallingFromPartitionOperationThread();
         invokeOnAllPartitions();
-
-        awaitCompletion();
-
-        retryFailedPartitions();
-
-        return (Map<Integer, T>) partitionResults;
+        return future;
     }
 
     private void ensureNotCallingFromPartitionOperationThread() {
+        // TODO [viliam] is this check necessary when it's async?
```

Comment:
Probably this check can be moved to a subclass of `SimpleCompletableFuture` with overridden `get(long, java.util.concurrent.TimeUnit)`.

Reply:
As per discussion we're leaving it here and remove the TODO.